<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="The function identifies  differentially methylated regions
in a case sample by comparing it against a control panel."><title>Epimutations analysis based on outlier detection methods — epimutations • epimutacions</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Epimutations analysis based on outlier detection methods — epimutations"><meta property="og:description" content="The function identifies  differentially methylated regions
in a case sample by comparing it against a control panel."><meta property="og:image" content="https://dpelegri.github.io/epimutacions/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">epimutacions</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/epimutacions.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="http://twitter.com/brgelab" aria-label="Twitter">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/isglobal-brge/epimutacions/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Epimutations analysis based on outlier detection methods</h1>
      <small class="dont-index">Source: <a href="https://github.com/isglobal-brge/epimutacions/blob/HEAD/R/epimutations.R" class="external-link"><code>R/epimutations.R</code></a></small>
      <div class="d-none name"><code>epimutations.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The function identifies  differentially methylated regions
in a case sample by comparing it against a control panel.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">epimutations</span><span class="op">(</span></span>
<span>  <span class="va">case_samples</span>,</span>
<span>  <span class="va">control_panel</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"manova"</span>,</span>
<span>  chr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  start <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  end <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  epi_params <span class="op">=</span> <span class="fu"><a href="epi_parameters.html">epi_parameters</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  maxGap <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  bump_cutoff <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  min_cpg <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>case_samples</dt>
<dd><p>a GenomicRatioSet object containing the case samples.
See the constructor function <a href="https://rdrr.io/pkg/minfi/man/GenomicRatioSet-class.html" class="external-link">GenomicRatioSet</a>,
<a href="https://rdrr.io/pkg/minfi/man/makeGenomicRatioSetFromMatrix.html" class="external-link">makeGenomicRatioSetFromMatrix</a>.</p></dd>


<dt>control_panel</dt>
<dd><p>a GenomicRatioSet object containing the
control panel (control panel).</p></dd>


<dt>method</dt>
<dd><p>a character string naming the
outlier detection method to be used.
This can be set as: <code>"manova"</code>,
<code>"mlm"</code>, <code>"iForest"</code>, <code>"mahdist"</code>,
<code>"quantile"</code> and <code>"beta"</code>.
The default is <code>"manova"</code>.
For more information see <strong>Details</strong>.</p></dd>


<dt>chr</dt>
<dd><p>a character string containing the sequence
names to be analysed. The default value is <code>NULL</code>.</p></dd>


<dt>start</dt>
<dd><p>an integer specifying the start position.
The default value is <code>NULL</code>.</p></dd>


<dt>end</dt>
<dd><p>an integer specifying the end position.
The default value is <code>NULL</code>.</p></dd>


<dt>epi_params</dt>
<dd><p>the parameters for each method.
See the function <a href="epi_parameters.html">epi_parameters</a>.</p></dd>


<dt>maxGap</dt>
<dd><p>the maximum location gap used in
<a href="https://rdrr.io/pkg/bumphunter/man/bumphunter.html" class="external-link">bumphunter</a> method.</p></dd>


<dt>bump_cutoff</dt>
<dd><p>a numeric value of the
estimate of the genomic profile above the
cutoff or below the negative of the
cutoff will be used as candidate regions.</p></dd>


<dt>min_cpg</dt>
<dd><p>an integer specifying the minimum CpGs number in a DMR.</p></dd>


<dt>verbose</dt>
<dd><p>logical. If TRUE additional details about
the procedure will provide to the user.
The default is TRUE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>The function returns an object of class tibble
containing the outliers regions.
The results are composed by the following columns:</p><ul><li><p><code>epi_id</code>: systematic name for each epimutation identified.
It provides the name of the used anomaly detection method.</p></li>
<li><p><code>sample</code>: the name of the sample containing the epimutation.</p></li>
<li><p><code>chromosome</code>, <code>start</code> and <code>end</code>:
indicate the location of the epimutation.</p></li>
<li><p><code>sz</code>: the window's size of the event.</p></li>
<li><p><code>cpg_n</code>: the number of CpGs in the epimutation.</p></li>
<li><p><code>cpg_ids</code>: the names of CpGs in the epimutation.</p></li>
<li><p><code>outlier_score</code>:</p><ul><li><p>For method <code>manova</code> it provides the approximation
to F-test and the Pillai score, separated by <code>/</code>.</p></li>
<li><p>For method <code>mlm</code> it provides the approximation to
F-test and the R2 of the model, separated by <code>/</code>.</p></li>
<li><p>For method <code>iForest</code> it provides
the magnitude of the outlier score.</p></li>
<li><p>For method <code>beta</code> it provides the mean outlier p-value.</p></li>
<li><p>For methods <code>quantile</code> and
<code>mahdist</code> it is filled with NA.</p></li>
</ul></li>
<li><p><code>outlier_direction</code>: indicates the direction
of the outlier with <code>"hypomethylation"</code> and <code>"hypermethylation"</code></p><ul><li><p>For <code>manova</code>, <code>mlm</code>, <code>iForest</code>, and <code>mahdist</code>
it is computed from the values obtained from bumphunter.</p></li>
<li><p>For <code>quantile</code> it is computed from the location
of the sample in the reference distribution (left vs. right outlier).</p></li>
<li><p>For method <code>beta</code> it return a NA.</p></li>
</ul></li>
<li><p><code>pvalue</code>:</p><ul><li><p>For methods <code>manova</code>, <code>mlm</code>, and <code>iForest</code>
it provides the p-value obtained from the model.</p></li>
<li><p>For method <code>quantile</code>, <code>mahdist</code> and <code>beta</code>
is filled with NA.</p></li>
</ul></li>
<li><p><code>adj_pvalue</code>: for methods with p-value (<code>manova</code> and
<code>mlm</code> adjusted p-value with Benjamini-Hochberg based on the total
number of regions detected by Bumphunter.</p></li>
<li><p><code>epi_region_id</code>: Name of the epimutation region as defined
in <code>candRegsGR</code>.</p></li>
<li><p><code>CRE</code>: cREs (cis-Regulatory Elements) as defined by ENCODE
overlapping the epimutation region. Different cREs are separated by ;.</p></li>
<li><p><code>CRE_type</code>: Type of cREs (cis-Regulatory Elements) as defined
by ENCODE. Different type are separeted by,
and different cREs are separated by ;.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function compares a case sample against
a control panel to identify epimutations in the given sample.
First, the DMRs are identified using the
<a href="https://rdrr.io/pkg/bumphunter/man/bumphunter.html" class="external-link">bumphunter</a> approach.
After that, CpGs in those DMRs are tested in order to detect regions
with CpGs being outliers.
For that, different outlier detection methods can be selected:</p><ul><li><p>Multivariate Analysis of Variance (<code>"manova"</code>). <a href="https://rdrr.io/r/stats/manova.html" class="external-link">manova</a></p></li>
<li><p>Multivariate Linear Model (<code>"mlm"</code>)</p></li>
<li><p>Isolation Forest (<code>"iForest"</code>) <a href="https://rdrr.io/pkg/isotree/man/isolation.forest.html" class="external-link">isolation.forest</a></p></li>
<li><p>Robust Mahalanobis Distance (<code>"mahdist"</code>)
<a href="https://rdrr.io/pkg/robustbase/man/covMcd.html" class="external-link">covMcd</a></p></li>
<li><p>Quantile distribution (<code>"quantile"</code>)</p></li>
<li><p>Beta (<code>"beta"</code>)</p></li>
</ul><p>We defined candidate epimutation regions (found in candRegsGR)
based on the 450K array design. As CpGs are not equally distributed
along the genome, only CpGs closer
to other CpGs can form an epimutation.
More information can be found in candRegsGR documentation.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">GRset</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#Find epimutations in GSM2562701 sample of GRset dataset</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">case_samples</span> <span class="op">&lt;-</span> <span class="va">GRset</span><span class="op">[</span>,<span class="fl">11</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">control_panel</span> <span class="op">&lt;-</span> <span class="va">GRset</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu">epimutations</span><span class="op">(</span><span class="va">case_samples</span>, <span class="va">control_panel</span>, method <span class="op">=</span> <span class="st">"manova"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Selected epimutation detection method 'manova'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Selected method 'manova' required of 'bumphunter'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [bumphunterEngine] Using a single core (backend: doSEQ, version: 1.5.2).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [bumphunterEngine] Computing coefficients.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [bumphunterEngine] Finding regions.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>NAs found and removed. ind changed.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> [bumphunterEngine] Found 120 bumps.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 9 candidate regions were found for case sample 'GSM2562701'</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> snapshotDate(): 2022-04-26</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 16</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   epi_id       sample  chrom…¹  start    end    sz cpg_n cpg_ids outli…² outli…³</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> epi_manova_1 GSM256… chr21   3.44<span style="color: #949494;">e</span>7 3.44<span style="color: #949494;">e</span>7   444     4 cg1417… 33.066… hyperm…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> epi_manova_2 GSM256… chr21   1.51<span style="color: #949494;">e</span>7 1.51<span style="color: #949494;">e</span>7   731     3 cg1229… 45.475… hypome…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> epi_manova_3 GSM256… chr21   1.54<span style="color: #949494;">e</span>7 1.54<span style="color: #949494;">e</span>7   428     3 cg1868… 60.011… hyperm…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> epi_manova_4 GSM256… chr21   4.70<span style="color: #949494;">e</span>7 4.70<span style="color: #949494;">e</span>7   535     4 cg2224… 11.926… hyperm…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> epi_manova_5 GSM256… chr21   1.51<span style="color: #949494;">e</span>7 1.51<span style="color: #949494;">e</span>7   523     3 cg0249… 44.043… hypome…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> epi_manova_6 GSM256… chr21   3.86<span style="color: #949494;">e</span>7 3.86<span style="color: #949494;">e</span>7   869     5 cg0512… 24.591… hyperm…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># … with 6 more variables: pvalue &lt;dbl&gt;, adj_pvalue &lt;dbl&gt;, delta_beta &lt;dbl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   epi_region_id &lt;lgl&gt;, CRE &lt;lgl&gt;, CRE_type &lt;lgl&gt;, and abbreviated variable</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   names ¹​chromosome, ²​outlier_score, ³​outlier_direction</span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Dolors Pelegri-Siso, Juan R. Gonzalez, Carlos Ruiz-Arenas, Carles Hernandez-Ferrer, Leire Abarrategui.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

