<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="epimutacions">
<title>The epimutacions User's Guide • epimutacions</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="The epimutacions User's Guide">
<meta property="og:description" content="epimutacions">
<meta property="og:image" content="https://dpelegri.github.io/epimutacions/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">epimutacions</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/epimutacions.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="http://twitter.com/brgelab" aria-label="Twitter">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/isglobal-brge/epimutacions/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="epimutacions_files/kePrint-0.0.1/kePrint.js"></script><link href="epimutacions_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>The epimutacions User's Guide</h1>
                        <h4 data-toc-skip class="author">Leire
Abarrategui</h4>
            <address class="author_afil">
      Bioinformatics Research Group in Epidemiology (BRGE)Barcelona
Insitute for Global Health (ISGlobal)Faculty of Medical Sciences,
Newcastle UniversityAutonomous University of Barcelona
(UAB)<br><a class="author_email" href="mailto:#"></a><a href="mailto:leire.abarrategui-martinez@newcastle.ac.uk" class="email">leire.abarrategui-martinez@newcastle.ac.uk</a>
      </address>
                              <h4 data-toc-skip class="author">Carles
Hernandez-Ferrer</h4>
            <address class="author_afil">
      Centro Nacional de Análisis Genómico (CNAG-CRG), Center for
Genomic RegulationBarcelona Institute of Science and Technology
(BIST)<br><a class="author_email" href="mailto:#"></a><a href="mailto:carles.hernandez@cnag.crg.eu" class="email">carles.hernandez@cnag.crg.eu</a>
      </address>
                              <h4 data-toc-skip class="author">Carlos
Ruiz-Arenas</h4>
            <address class="author_afil">
      Centro de Investigación Biomédica en Red de Enfermedades Raras
(CIBERER)Universitat Pompeu Fabra
(UPF)<br><a class="author_email" href="mailto:#"></a><a href="mailto:carlos.ruiza@upf.edu" class="email">carlos.ruiza@upf.edu</a>
      </address>
                              <h4 data-toc-skip class="author">Juan R.
Gonzalez</h4>
            <address class="author_afil">
      Bioinformatics Research Group in Epidemiology (BRGE),Barcelona
Insitute for Global Health (ISGlobal)Department of Mathematics,
Autonomous University of Barcelona
(UAB)<br><a class="author_email" href="mailto:#"></a><a href="mailto:juanr.gonzalez@isglobal.org" class="email">juanr.gonzalez@isglobal.org</a>
      </address>
                  
            <h4 data-toc-skip class="date">2022-10-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/isglobal-brge/epimutacions/blob/HEAD/vignettes/epimutacions.Rmd" class="external-link"><code>vignettes/epimutacions.Rmd</code></a></small>
      <div class="d-none name"><code>epimutacions.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Epimutations are rare alterations in the methylation pattern at
      specific loci. Have been demonstrated that epimutations could be
      the causative factor of some genetic diseases. Nonetheless, no
      standard methods are available to detect and quantify these
      alterations. This vignette provides an introductory guide to the
      <code>epimutacions</code> package. The package contains several
      outlier detection methods to identify epimutations in genome-wide
      DNA methylation microarrays data. The areas covered in this
      document are: (1) package installation; (2) data loading and
      preprocessing; and (3) epimutation identification, annotation and
      visualization.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="section level3">
<h3 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h3>
<p>Epimutations are rare alterations in the methylation pattern at
specific loci. Have been demonstrated that epimutations could be the
causative factor of some genetic diseases. For example, epimutations can
lead to cancers, such as Lynch syndrome, rare diseases such as
Prader-Willi syndrome, and are associated with common disorders, such as
autism. Nonetheless, no standard methods are available to detect and
quantify these alterations. Two methods for epimutations detection on
methylation microarray data have been reported: (1) based on identifying
CpGs with outlier values and then cluster them in epimutations <span class="citation">(Barbosa et al. 2018)</span>; (2) define candidate
regions with bumphunter and test their statistical significance with a
MANOVA <span class="citation">(Aref-Eshghi et al. 2019)</span>. However,
the implementation of these methods is not publicly available, and these
approaches have not been compared.</p>
<p>We have developed <code>epimutacions</code> R/Bioconductor package.
We have implemented those two methods (called <code>quantile</code> and
<code>manova</code>, respectively). We implemented four additional
approaches, using a different distribution to detect CpG outliers
(<code>beta</code>), or a different model to assess region significance
(<code>mlm</code>, <code>mahdist</code>, and <code>iForest</code>).</p>
<p>The package <code>epimutacions</code> provides tools to raw DNA
methylation microarray intensities normalization and epimutations
identification, visualization and annotation.<br>
The full <code>epimutacions</code> user´s guide is available in this
vignette.<br>
The main function to estimate the epimutations is called
<code><a href="../reference/epimutations.html">epimutations()</a></code>.</p>
<p>The name of the package is <code>epimutacions</code> (pronounced
<code>ɛ pi mu ta 'sj ons</code>) which means epimutations in Catalan, a
language from the northeast of Spain.</p>
</div>
<div class="section level3">
<h3 id="methodology">Methodology<a class="anchor" aria-label="anchor" href="#methodology"></a>
</h3>
<p>The <code>epimutacions</code> package computes a genome-wide DNA
methylation analysis to identify the epimutations to be considered as
biomarkers for case samples with a suspected genetic disease. The
function <code><a href="../reference/epimutations.html">epimutations()</a></code> infers epimutations on a
case-control design. It compares a case sample against a reference panel
(healthy individuals). The package also includes leave-one-out approach
(<code><a href="../reference/epimutations_one_leave_out.html">epimutations_one_leave_out()</a></code>). It compares individual
methylation profiles of a single sample (regardless if they are cases or
controls) against all other samples from the same cohort.</p>
<p>The <code>epimutacions</code> package includes 6 outlier detection
approaches (figure @ref(fig:implementation)): (1) Multivariate Analysis
of variance (<code>manova</code>), (2) Multivariate Linear Model
(<code>mlm</code>), (3) isolation forest (<code>iForest</code>), (4)
robust mahalanobis distance (<code>mahdist</code>) (5)
<code>quantile</code> and (6) <code>beta</code>.</p>
<p>The approaches <code>manova</code>, <code>mlm</code>,
<code>iForest</code> and <code>mahdist</code> define the candidate
regions (differentially methylated regions (DMRs)) using bumphunter
method <span class="citation">(Jaffe et al. 2012)</span>. Then, those
DMRs are tested to identify regions with CpGs being outliers when
comparing with the reference panel. <code>quantile</code> uses the
sliding window approach to individually compare the methylation value in
each proband against the reference panel and then cluster them in
epimutations.<br><code>Beta</code> utilizes beta distribution to identify outlier
CpGs.</p>
<p>We have defined an epimutation as a consecutive window of a minimum
of 3 outliers CpGs with a maximum distance of 1kb between them <span class="citation">(Barbosa et al. 2018)</span>.</p>
<div class="figure" style="text-align: center">
<img src="fig/implementation.png" alt="Implementation of each outlier detection method" width="90%"><p class="caption">
Implementation of each outlier detection method
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="section level3">
<h3 id="installing-the-packages">Installing the packages<a class="anchor" aria-label="anchor" href="#installing-the-packages"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"epimutacions"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"epimutacionsData"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="loading-libraries">Loading libraries<a class="anchor" aria-label="anchor" href="#loading-libraries"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/isglobal-brge/epimutacions" class="external-link">epimutacions</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h3>
<p>The workflow in figure @ref(fig:workflow) explains the main analysis
in the <code>epimutacions</code> package.</p>
<p>The package allows two different types of inputs:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Case samples <code>IDAT</code> files (raw microarray intensities)
together with <code>RGChannelSet</code> object as reference panel. The
reference panel can be supplied by the user or can be selected through
the example datasets that the package provides (section
@ref(datasets)).</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>
<code>GenomicRatioSet</code> object containing case and control
samples.</li>
</ol></li>
</ul>
<p>The normalization (<code><a href="../reference/epi_preprocess.html">epi_preprocess()</a></code>) converts the raw
microarray intensities into usable methylation measurement (<span class="math inline">\(\beta\)</span> values at CpG locus level). As a
result, we obtain a <code>GenomicRatioSet</code> object, which can be
used as <code><a href="../reference/epimutations.html">epimutations()</a></code> function input. The data should
contain information about values of CpG sites, phenotype and feature
data.</p>
<div class="figure" style="text-align: center">
<img src="fig/workflow.png" alt="Allowed data formats, normalization and input types" width="90%"><p class="caption">
Allowed data formats, normalization and input types
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="datasets">Datasets<a class="anchor" aria-label="anchor" href="#datasets"></a>
</h2>
<p>The package contains 3 example datasets adapted from <a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">Gene Expression Omnibus
(GEO)</a>:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>4 case samples IDAT files <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE131350" class="external-link">(GEO:
GSE131350)</a>
</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>
<code>reference_panel</code>: a <code>RGChannelSet</code> class
object containing 22 healthy individuals <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE12782" class="external-link">(GEO:
GSE127824)</a>
</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>
<code>methy</code>: a <code>GenomicRatioSet</code> object which
includes 49 controls <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE104812" class="external-link">(GEO:
GSE104812)</a> and 3 cases <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi" class="external-link">(GEO:
GSE97362)</a>.</li>
</ol></li>
</ul>
<p>We also included a dataset specifying the 40,408 candidate regions in
Illumina 450K array which could be epimutations (see section
@ref(candreg)).</p>
<p>We created the <code>epimutacionsData</code> package in
<code>ExperimentHub</code>. It contains the reference panel, methy and
the candidate epimutations datasets. The package includes the IDAT files
as external data. To access the datasets we need to install the packages
by running the following commands:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ExperimentHub"</span><span class="op">)</span></span></code></pre></div>
<p>Then, we need to load the package and create an
<code>ExperimentHub</code> object:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">eh</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"epimutacionsData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>ExperimentHub with 3 records
# snapshotDate(): 2022-04-26
# $dataprovider: GEO, Illumina 450k array
# $species: Homo sapiens
# $rdataclass: RGChannelSet, GenomicRatioSet, GRanges
# additional mcols(): taxonomyid, genome, description,
#   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
#   rdatapath, sourceurl, sourcetype 
# retrieve records with, e.g., 'object[["EH6690"]]' 

           title                   
  EH6690 | Control and case samples
  EH6691 | Reference panel         
  EH6692 | Candidate epimutations  </code></pre>
<div class="section level3">
<h3 id="idat-files-and-reference-panel">IDAT files and reference panel<a class="anchor" aria-label="anchor" href="#idat-files-and-reference-panel"></a>
</h3>
<p>IDAT files directory in <code>epimutacionsData</code> package:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">baseDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"epimutacionsData"</span><span class="op">)</span></span></code></pre></div>
<p>The reference panel and <code>methy</code> dataset can be found in
<code>EH6691</code> and <code>EH6690</code> record of the
<code>eh</code> object:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reference_panel</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH6691"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">methy</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH6690"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="candreg">Candidate regions<a class="anchor" aria-label="anchor" href="#candreg"></a>
</h3>
<p>In Illumina 450K array, probes are unequally distributed along the
genome, limiting the number of regions that can fulfil the requirements
to be considered an epimutation. Thus, we have computed a dataset
containing all the regions that could be candidates to become an
epimutation.</p>
<p>We used the clustering approach from bumphunter to define the
candidate epimutations. We defined a primary dataset with all the CpGs
from the Illumina 450K array. Then, we run bumphunter and selected those
regions with at least 3 CpGs and a maximum distance of 1kb between them.
As a result, we found 40,408 candidate epimutations. The function
<code>epimutation()</code> filters the identified epimutations using
these candidate regions.</p>
<p>The following is the code used to identify the candidate epimutations
in Illumina 450K array:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hansenlab/minfi" class="external-link">minfi</a></span><span class="op">)</span></span>
<span><span class="co"># Regions 450K</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"IlluminaHumanMethylation450kanno.ilmn12.hg19"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Locations</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Select CpGs (names starting by cg) in autosomic chromosomes</span></span>
<span><span class="va">locs.450</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">Locations</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^cg"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">Locations</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">chr</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">locs.450GR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html" class="external-link">makeGRangesFromDataFrame</a></span><span class="op">(</span><span class="va">locs.450</span>, </span>
<span>                                       start.field <span class="op">=</span> <span class="st">"pos"</span>, </span>
<span>                                       end.field <span class="op">=</span> <span class="st">"pos"</span>, </span>
<span>                                       strand <span class="op">=</span> <span class="st">"*"</span><span class="op">)</span></span>
<span><span class="va">locs.450GR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">locs.450GR</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">locs.450GR</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span>, </span>
<span>              dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">locs.450GR</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Set sample B to all 1</span></span>
<span><span class="va">mat</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="co">## Define model matrix</span></span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">var</span>, <span class="va">pheno</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Run bumphunter</span></span>
<span><span class="va">bumps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bumphunter/man/bumphunter.html" class="external-link">bumphunter</a></span><span class="op">(</span><span class="va">mat</span>, design <span class="op">=</span> <span class="va">model</span>, pos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">locs.450GR</span><span class="op">)</span>, </span>
<span>                    chr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">locs.450GR</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                    cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">$</span><span class="va">table</span></span>
<span><span class="va">bumps.fil</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">bumps</span>, <span class="va">L</span> <span class="op">&gt;=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p>The candidate regions can be found in <code>EH6692</code> record of
the <code>eh</code> object:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">candRegsGR</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH6692"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<p>The <code><a href="../reference/epi_preprocess.html">epi_preprocess()</a></code> function allows calling the 6
preprocessing methods from <code>minfi</code> package:</p>
<table class="table">
<colgroup>
<col width="13%">
<col width="25%">
<col width="60%">
</colgroup>
<thead><tr class="header">
<th>Method</th>
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>raw</code></td>
<td><code>preprocessRaw</code></td>
<td>Converts the Red/Green channel for an Illumina methylation array
into methylation signal, without using any normalization</td>
</tr>
<tr class="even">
<td><code>illumina</code></td>
<td><code>preprocessIllumina</code></td>
<td>Implements preprocessing for Illumina methylation microarrays as
used in Genome Studio</td>
</tr>
<tr class="odd">
<td><code>swan</code></td>
<td><code>preprocessSWAN</code></td>
<td>Subset-quantile Within Array Normalisation (SWAN). It allows
Infinium I and II type probes on a single array to be normalized
together</td>
</tr>
<tr class="even">
<td><code>quantile</code></td>
<td><code>preprocessQuantile</code></td>
<td>Implements stratified quantile normalization preprocessing for
Illumina methylation microarrays</td>
</tr>
<tr class="odd">
<td><code>noob</code></td>
<td><code>preprocessNoob</code></td>
<td>Noob (normal-exponential out-of-band) is a background correction
method with dye-bias normalization for Illumina Infinium methylation
arrays</td>
</tr>
<tr class="even">
<td><code>funnorm</code></td>
<td><code>preprocessFunnorm</code></td>
<td>Functional normalization (FunNorm) is a between-array normalization
method for the Illumina Infinium HumanMethylation450 platform</td>
</tr>
</tbody>
</table>
<p>Each normalization approach has some unique parameters which can be
modified through <code><a href="../reference/norm_parameters.html">norm_parameters()</a></code> function:</p>
<table class="table">
<thead><tr>
<th style="text-align:left;">
parameters
</th>
<th style="text-align:left;">
Description
</th>
</tr></thead>
<tbody>
<tr grouplength="3">
<td colspan="2" style="border-bottom: 1px solid;">
<strong>illumina</strong>
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
bg.correct
</td>
<td style="text-align:left;">
Performs background correction
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
normalize
</td>
<td style="text-align:left;">
Performs controls normalization
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
reference
</td>
<td style="text-align:left;">
The reference array for control normalization
</td>
</tr>
<tr grouplength="7">
<td colspan="2" style="border-bottom: 1px solid;">
<strong>quantile</strong>
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
fixOutliers
</td>
<td style="text-align:left;">
Low outlier Meth and Unmeth signals will be fixed
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
removeBadSamples
</td>
<td style="text-align:left;">
Remove bad samples
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
badSampleCutoff
</td>
<td style="text-align:left;">
The cutoff to label samples as ‘bad’
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
quantileNormalize
</td>
<td style="text-align:left;">
Performs quantile normalization
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
stratified
</td>
<td style="text-align:left;">
Performs quantile normalization within region strata
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
mergeManifest
</td>
<td style="text-align:left;">
Merged to the output the information in the associated manifest package
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
sex
</td>
<td style="text-align:left;">
Sex of the samples
</td>
</tr>
<tr grouplength="3">
<td colspan="2" style="border-bottom: 1px solid;">
<strong>noob</strong>
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
offset
</td>
<td style="text-align:left;">
Offset for the normexp background correct
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
dyeCorr
</td>
<td style="text-align:left;">
Performs dye normalization
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
dyeMethod
</td>
<td style="text-align:left;">
Dye bias correction to be done
</td>
</tr>
<tr grouplength="5">
<td colspan="2" style="border-bottom: 1px solid;">
<strong>funnorm</strong>
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
nPCs
</td>
<td style="text-align:left;">
The number of principal components from the control probes
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
sex
</td>
<td style="text-align:left;">
Sex of the samples
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
bgCorr
</td>
<td style="text-align:left;">
Performs NOOB background correction before functional normalization
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
dyeCorr
</td>
<td style="text-align:left;">
Performs dye normalization
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
keepCN
</td>
<td style="text-align:left;">
Keeps copy number estimates
</td>
</tr>
</tbody>
</table>
<p>We can obtain the default settings for each method by invoking the
function <code><a href="../reference/norm_parameters.html">norm_parameters()</a></code> with no arguments:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/norm_parameters.html">norm_parameters</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code>$illumina
$illumina$bg.correct
[1] TRUE

$illumina$normalize
[1] "controls" "no"      

$illumina$reference
[1] 1


$quantile
$quantile$fixOutliers
[1] TRUE

$quantile$removeBadSamples
[1] FALSE

$quantile$badSampleCutoff
[1] 10.5

$quantile$quantileNormalize
[1] TRUE

$quantile$stratified
[1] TRUE

$quantile$mergeManifest
[1] FALSE

$quantile$sex
NULL


$noob
$noob$offset
[1] 15

$noob$dyeCorr
[1] TRUE

$noob$dyeMethod
[1] "single"    "reference"


$funnorm
$funnorm$nPCs
[1] 2

$funnorm$sex
NULL

$funnorm$bgCorr
[1] TRUE

$funnorm$dyeCorr
[1] TRUE

$funnorm$keepCN
[1] FALSE</code></pre>
<p>However, we can modify the parameter(s) as the following example for
<code>illumina</code> approach:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/norm_parameters.html">norm_parameters</a></span><span class="op">(</span>illumina <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"bg.correct"</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">parameters</span><span class="op">$</span><span class="va">illumina</span><span class="op">$</span><span class="va">bg.correct</span></span></code></pre></div>
<pre><code>[1] FALSE</code></pre>
<p>We are going to preprocess the IDAT files and reference panel
(@ref(datasets)). We need to specify the IDAT files directory and the
reference panel in <code>RGChannelSet</code> format. As a result, we
will obtain a <code>GenomicRatioSet</code> object containing the control
and case samples:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GRset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/epi_preprocess.html">epi_preprocess</a></span><span class="op">(</span><span class="va">baseDir</span>, </span>
<span>                        <span class="va">reference_panel</span>, </span>
<span>                        pattern <span class="op">=</span> <span class="st">"SampleSheet.csv"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="epimutations">Epimutations<a class="anchor" aria-label="anchor" href="#epimutations"></a>
</h2>
<div class="section level3">
<h3 id="epimutations-detection">Epimutations detection<a class="anchor" aria-label="anchor" href="#epimutations-detection"></a>
</h3>
<p>The function <code><a href="../reference/epimutations.html">epimutations()</a></code> includes 6 methods for
epimutation identification: (1) Multivariate Analysis of variance
(<code>manova</code>), (2) Multivariate Linear Model (<code>mlm</code>),
(3) isolation forest (<code>iForest</code>), (4) robust mahalanobis
distance (<code>mahdist</code>) (5) <code>quantile</code> and (6)
<code>beta</code>.</p>
<p>To illustrate the following examples we are going to use the dataset
<code>methy</code> (section @ref(methy)). We need to separate the case
and control samples:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">case_samples</span> <span class="op">&lt;-</span> <span class="va">methy</span><span class="op">[</span>,<span class="va">methy</span><span class="op">$</span><span class="va">status</span> <span class="op">==</span> <span class="st">"case"</span><span class="op">]</span></span>
<span><span class="va">control_samples</span> <span class="op">&lt;-</span> <span class="va">methy</span><span class="op">[</span>,<span class="va">methy</span><span class="op">$</span><span class="va">status</span> <span class="op">==</span> <span class="st">"control"</span><span class="op">]</span></span></code></pre></div>
<p>We can specify the chromosome or region to analyze which helps to
reduce the execution time:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">epi_mvo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/epimutations.html">epimutations</a></span><span class="op">(</span><span class="va">case_samples</span>, </span>
<span>                        <span class="va">control_samples</span>, </span>
<span>                        method <span class="op">=</span> <span class="st">"manova"</span><span class="op">)</span></span></code></pre></div>
<p>The function <code><a href="../reference/epimutations_one_leave_out.html">epimutations_one_leave_out()</a></code> compared
individual methylation profiles of a single sample (regardless if they
are cases or controls) against all other samples from the same cohort.
To use this function we do not need to split the dataset. To ilustrate
this example we are going to use the <code>GRset</code> dataset
available in <code>epimutacions</code> package:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#manova (default method)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">GRset</span><span class="op">)</span></span>
<span><span class="va">epi_mva_one_leave_out</span><span class="op">&lt;-</span> <span class="fu"><a href="../reference/epimutations_one_leave_out.html">epimutations_one_leave_out</a></span><span class="op">(</span><span class="va">GRset</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="unique-parameters">Unique parameters<a class="anchor" aria-label="anchor" href="#unique-parameters"></a>
</h3>
<p>The <code><a href="../reference/epi_parameters.html">epi_parameters()</a></code> function is useful to set the
individual parameters for each outliers detection approach. The
following table describes the arguments:</p>
<table class="table">
<thead><tr>
<th style="text-align:left;">
parameters
</th>
<th style="text-align:left;">
Description
</th>
</tr></thead>
<tbody>
<tr grouplength="1">
<td colspan="2" style="border-bottom: 1px solid;">
<strong>Manova, mlm</strong>
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
pvalue_cutoff
</td>
<td style="text-align:left;">
The threshold p-value to select which CpG regions are outliers
</td>
</tr>
<tr grouplength="2">
<td colspan="2" style="border-bottom: 1px solid;">
<strong>iso.forest</strong>
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
outlier_score_cutoff
</td>
<td style="text-align:left;">
The threshold to select which CpG regions are outliers
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
ntrees
</td>
<td style="text-align:left;">
The number of binary trees to build for the model
</td>
</tr>
<tr grouplength="1">
<td colspan="2" style="border-bottom: 1px solid;">
<strong>mahdist.mcd</strong>
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
nsamp
</td>
<td style="text-align:left;">
The number of subsets used for initial estimates in the MCD
</td>
</tr>
<tr grouplength="2">
<td colspan="2" style="border-bottom: 1px solid;">
<strong>quantile</strong>
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
window_sz
</td>
<td style="text-align:left;">
The maximum distance between CpGs to be considered in the same DMR
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
offset_mean/offset_abs
</td>
<td style="text-align:left;">
The upper and lower threshold to consider a CpG an outlier
</td>
</tr>
<tr grouplength="2">
<td colspan="2" style="border-bottom: 1px solid;">
<strong>beta</strong>
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
pvalue_cutoff
</td>
<td style="text-align:left;">
The minimum p-value to consider a CpG an outlier
</td>
</tr>
<tr>
<td style="text-align:left;padding-left: 2em;" indentlevel="1">
diff_threshold
</td>
<td style="text-align:left;">
The minimum methylation difference between the CpG and the mean
methylation to consider a position an outlier
</td>
</tr>
</tbody>
</table>
<p><code><a href="../reference/epi_parameters.html">epi_parameters()</a></code> with no arguments, returns a list of
the default settings for each method:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/epi_parameters.html">epi_parameters</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code>$manova
$manova$pvalue_cutoff
[1] 0.05


$mlm
$mlm$pvalue_cutoff
[1] 0.05


$iForest
$iForest$outlier_score_cutoff
[1] 0.7

$iForest$ntrees
[1] 100


$mahdist
$mahdist$nsamp
[1] "deterministic"


$quantile
$quantile$window_sz
[1] 1000

$quantile$offset_abs
[1] 0.15

$quantile$qsup
[1] 0.995

$quantile$qinf
[1] 0.005


$beta
$beta$pvalue_cutoff
[1] 1e-06

$beta$diff_threshold
[1] 0.1</code></pre>
<p>The set up of any parameter can be done as the following example for
<code>manova</code>:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/epi_parameters.html">epi_parameters</a></span><span class="op">(</span>manova <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"pvalue_cutoff"</span> <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">parameters</span><span class="op">$</span><span class="va">manova</span><span class="op">$</span><span class="va">pvalue_cutoff</span></span></code></pre></div>
<pre><code>[1] 0.01</code></pre>
</div>
<div class="section level3">
<h3 id="results-description">Results description<a class="anchor" aria-label="anchor" href="#results-description"></a>
</h3>
<p>The <code>epimutations</code> function returns a data frame (tibble)
containing all the epimutations identified in the given case sample. If
no epimutation is found, the function returns a row containing the case
sample information and missing values for all other arguments. The
following table describes each argument:</p>
<table class="table">
<colgroup>
<col width="16%">
<col width="83%">
</colgroup>
<thead><tr class="header">
<th>Column name</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>epi_id</code></td>
<td>systematic name for each epimutation identified</td>
</tr>
<tr class="even">
<td><code>sample</code></td>
<td>The name of the sample containing that epimutation</td>
</tr>
<tr class="odd">
<td>
<code>chromosome</code> <code>start</code> <code>end</code>
</td>
<td>The location of the epimutation</td>
</tr>
<tr class="even">
<td><code>sz</code></td>
<td>The window’s size of the event</td>
</tr>
<tr class="odd">
<td><code>cpg_ids</code></td>
<td>The number of CpGs in the epimutation</td>
</tr>
<tr class="even">
<td><code>cpg_n</code></td>
<td>The names of CpGs in the epimutation</td>
</tr>
<tr class="odd">
<td><code>outlier_score</code></td>
<td>For method <code>manova</code> it provides the approximation to
F-test and the Pillai score, separated by <code>/</code><br> For method
<code>mlm</code> it provides the approximation to F-test and the R2 of
the model, separated by <code>/</code><br> For method
<code>iForest</code> it provides the magnitude of the outlier score.<br>
For method <code>beta</code> it provides the mean p-value of all GpGs in
that DMR<br> For methods <code>quantile</code> and <code>mahdist</code>
it is filled with <code>NA</code>.</td>
</tr>
<tr class="even">
<td><code>pvalue</code></td>
<td>For methods <code>manova</code> and <code>mlm</code> it provides the
p-value obtained from the model.<br> For method <code>quantile</code>,
<code>iForest</code>, <code>beta</code> and <code>mahdist</code> it is
filled with <code>NA</code>.</td>
</tr>
<tr class="odd">
<td><code>outlier_direction</code></td>
<td>Indicates the direction of the outlier with “hypomethylation” and
“hypermethylation”.<br> For <code>manova</code>, <code>mlm</code>,
<code>iForest</code>, and <code>mahdist</code> it is computed from the
values obtained from <code>bumphunter</code>.<br> For <code>beta</code>
is computed from the p value for each CpG using
<code>diff_threshold</code> and <code>pvalue_threshold</code>
arguments.<br> For <code>quantile</code> it is computed from the
location of the sample in the reference distribution (left vs. right
outlier).</td>
</tr>
<tr class="even">
<td><code>adj_pvalue</code></td>
<td>For methods <code>manova</code> and <code>mlm</code> it provides the
adjusted p-value with Benjamini-Hochberg based on the total number of
regions detected by Bumphunter.<br> For method <code>quantile</code>,
<code>iForest</code>, <code>mahdist</code> and <code>beta</code> it is
filled with <code>NA</code>.</td>
</tr>
<tr class="odd">
<td><code>epi_region_id</code></td>
<td>Name of the epimutation region as defined in
<code>candRegsGR</code>.</td>
</tr>
<tr class="even">
<td><code>CRE</code></td>
<td>cREs (cis-Regulatory Elements) as defined by ENCODE overlapping the
epimutation region.</td>
</tr>
<tr class="odd">
<td><code>CRE_type</code></td>
<td>Type of cREs (cis-Regulatory Elements) as defined by ENCODE.</td>
</tr>
</tbody>
</table>
<p>As an example we are going to visualize the obtained results with
MANOVA method (<code>epi_mvo</code>):</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">epi_mvo</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] 51 16</code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">epi_mvo</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] "tbl_df"     "tbl"        "data.frame"</code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">epi_mvo</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code>        epi_id     sample chromosome    start      end  sz cpg_n
1 epi_manova_1 GSM2562699      chr19 12777736 12777903 167     7
                                                                       cpg_ids
1 cg20791841,cg25267526,cg03641858,cg25441478,cg14132016,cg23954461,cg03143365
                       outlier_score outlier_direction       pvalue
1 302.558383959434/0.981008923520811  hypermethylation 3.441085e-33
    adj_pvalue delta_beta  epi_region_id                                    CRE
1 2.236705e-31  0.2960902 chr19_12776725 EH38E1939817,EH38E1939818,EH38E1939819
                             CRE_type
1 pELS;pELS,CTCF-bound;PLS,CTCF-bound</code></pre>
</div>
<div class="section level3">
<h3 id="epimutations-annotations">Epimutations annotations<a class="anchor" aria-label="anchor" href="#epimutations-annotations"></a>
</h3>
<p>The <code><a href="../reference/annotate_epimutations.html">annotate_epimutations()</a></code> function enriches the
identified epimutations. It includes information about GENECODE gene
names, description of the regulatory feature provided by methylation
consortium, the location of the CpG relative to the CpG island, OMIM
accession and description number and Ensembl region id, coordinates,
type and tissue:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rst_mvo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotate_epimutations.html">annotate_epimutations</a></span><span class="op">(</span><span class="va">epi_mvo</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="23%">
<col width="76%">
</colgroup>
<thead><tr class="header">
<th>Column name</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>GencodeBasicV12_NAME</code></td>
<td>Gene names from the basic GENECODE build</td>
</tr>
<tr class="even">
<td><code>Regulatory_Feature_Group</code></td>
<td>Description of the regulatory feature provided by the Methylation
Consortium</td>
</tr>
<tr class="odd">
<td><code>Relation_to_Island</code></td>
<td>The location of the CpG relative to the CpG island</td>
</tr>
<tr class="even">
<td>
<code>OMIM_ACC</code><br><code>OMIM_DESC</code>
</td>
<td>OMIM accession and description number</td>
</tr>
<tr class="odd">
<td>
<code>ensembl_reg_id</code><br><code>ensembl_reg_coordinates</code><br><code>ensembl_reg_type</code><br><code>ensembl_reg_tissues</code>
</td>
<td>The Ensembl region id, coordinates, type and tissue</td>
</tr>
</tbody>
</table>
<p>As an example we are going to visualize different annotations</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">kableExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span></span>
<span>    <span class="va">rst_mvo</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">27</span>,<span class="fl">32</span><span class="op">)</span> ,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"epi_id"</span>, <span class="st">"cpg_ids"</span>, <span class="st">"Relation_to_Island"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>    row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">column_spec</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, width <span class="op">=</span> <span class="st">"4cm"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
epi_id
</th>
<th style="text-align:left;">
cpg_ids
</th>
<th style="text-align:left;">
Relation_to_Island
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;width: 4cm; ">
epi_manova_44
</td>
<td style="text-align:left;width: 4cm; ">
cg19560927,cg02876326,cg16167052
</td>
<td style="text-align:left;width: 4cm; ">
N_Shore/N_Shore/Island
</td>
</tr>
<tr>
<td style="text-align:left;width: 4cm; ">
epi_manova_51
</td>
<td style="text-align:left;width: 4cm; ">
cg01396855,cg08684893,cg02386644
</td>
<td style="text-align:left;width: 4cm; ">
N_Shore/N_Shore/N_Shore
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">kableExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span></span>
<span>    <span class="va">rst_mvo</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">8</span>,<span class="fl">22</span><span class="op">)</span> , <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"epi_id"</span>, <span class="st">"OMIM_ACC"</span>, <span class="st">"OMIM_DESC"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>    row.names <span class="op">=</span> <span class="cn">FALSE</span> <span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">column_spec</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, width <span class="op">=</span> <span class="st">"4cm"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
epi_id
</th>
<th style="text-align:left;">
OMIM_ACC
</th>
<th style="text-align:left;">
OMIM_DESC
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;width: 4cm; ">
epi_manova_4
</td>
<td style="text-align:left;width: 4cm; ">
248500
</td>
<td style="text-align:left;width: 4cm; ">
MANNOSIDOSIS, ALPHA B, LYSOSOMAL; MANSA;;ALPHA-MANNOSIDOSIS;;LYSOSOMAL
ALPHA-D-MANNOSIDASE DEFICIENCY;;ALPHA-MANNOSIDASE B DEFICIENCY
</td>
</tr>
<tr>
<td style="text-align:left;width: 4cm; ">
epi_manova_12
</td>
<td style="text-align:left;width: 4cm; ">
616689/616689/616689/616689
</td>
<td style="text-align:left;width: 4cm; ">
DEHYDRATED HEREDITARY STOMATOCYTOSIS 2; DHS2;;XEROCYTOSIS
GARDOS;;DESICCYTOSIS GARDOS/DEHYDRATED HEREDITARY STOMATOCYTOSIS 2;
DHS2;;XEROCYTOSIS GARDOS;;DESICCYTOSIS GARDOS/DEHYDRATED HEREDITARY
STOMATOCYTOSIS 2; DHS2;;XEROCYTOSIS GARDOS;;DESICCYTOSIS
GARDOS/DEHYDRATED HEREDITARY STOMATOCYTOSIS 2; DHS2;;XEROCYTOSIS
GARDOS;;DESICCYTOSIS GARDOS
</td>
</tr>
<tr>
<td style="text-align:left;width: 4cm; ">
epi_manova_33
</td>
<td style="text-align:left;width: 4cm; ">
231090/231090/231090/231090/231090/231090
</td>
<td style="text-align:left;width: 4cm; ">
HYDATIDIFORM MOLE, RECURRENT, 1; HYDM1;;HYDATIDIFORM MOLE;
HYDM;;HYDATIDIFORM MOLE, COMPLETE; CHM;;GESTATIONAL TROPHOBLASTIC
DISEASE/HYDATIDIFORM MOLE, RECURRENT, 1; HYDM1;;HYDATIDIFORM MOLE;
HYDM;;HYDATIDIFORM MOLE, COMPLETE; CHM;;GESTATIONAL TROPHOBLASTIC
DISEASE/HYDATIDIFORM MOLE, RECURRENT, 1; HYDM1;;HYDATIDIFORM MOLE;
HYDM;;HYDATIDIFORM MOLE, COMPLETE; CHM;;GESTATIONAL TROPHOBLASTIC
DISEASE/HYDATIDIFORM MOLE, RECURRENT, 1; HYDM1;;HYDATIDIFORM MOLE;
HYDM;;HYDATIDIFORM MOLE, COMPLETE; CHM;;GESTATIONAL TROPHOBLASTIC
DISEASE/HYDATIDIFORM MOLE, RECURRENT, 1; HYDM1;;HYDATIDIFORM MOLE;
HYDM;;HYDATIDIFORM MOLE, COMPLETE; CHM;;GESTATIONAL TROPHOBLASTIC
DISEASE/HYDATIDIFORM MOLE, RECURRENT, 1; HYDM1;;HYDATIDIFORM MOLE;
HYDM;;HYDATIDIFORM MOLE, COMPLETE; CHM;;GESTATIONAL TROPHOBLASTIC
DISEASE
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">kableExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span></span>
<span>    <span class="va">rst_mvo</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"epi_id"</span>, <span class="st">"ensembl_reg_id"</span>, <span class="st">"ensembl_reg_type"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>    row.names <span class="op">=</span> <span class="cn">FALSE</span> <span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">column_spec</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, width <span class="op">=</span> <span class="st">"4cm"</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
epi_id
</th>
<th style="text-align:left;">
ensembl_reg_id
</th>
<th style="text-align:left;">
ensembl_reg_type
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;width: 4cm; ">
epi_manova_1
</td>
<td style="text-align:left;width: 4cm; ">
ENSR00000287480
</td>
<td style="text-align:left;width: 4cm; ">
Promoter
</td>
</tr>
<tr>
<td style="text-align:left;width: 4cm; ">
epi_manova_2
</td>
<td style="text-align:left;width: 4cm; ">
ENSR00000215007
</td>
<td style="text-align:left;width: 4cm; ">
Promoter
</td>
</tr>
<tr>
<td style="text-align:left;width: 4cm; ">
epi_manova_3
</td>
<td style="text-align:left;width: 4cm; ">
ENSR00000325316
</td>
<td style="text-align:left;width: 4cm; ">
Promoter
</td>
</tr>
<tr>
<td style="text-align:left;width: 4cm; ">
epi_manova_4
</td>
<td style="text-align:left;width: 4cm; ">
ENSR00001156259///ENSR00001156260
</td>
<td style="text-align:left;width: 4cm; ">
Promoter///CTCF Binding Site
</td>
</tr>
<tr>
<td style="text-align:left;width: 4cm; ">
epi_manova_5
</td>
<td style="text-align:left;width: 4cm; ">
ENSR00000324412
</td>
<td style="text-align:left;width: 4cm; ">
CTCF Binding Site
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="epimutation-visualization">Epimutation visualization<a class="anchor" aria-label="anchor" href="#epimutation-visualization"></a>
</h3>
<p>The <code><a href="../reference/plot_epimutations.html">plot_epimutations()</a></code> function locates the
epimutations along the genome. It plots the methylation values of the
case sample in red, the control samples in dashed black lines and
population mean in blue:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_epimutations.html">plot_epimutations</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">epi_mvo</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span><span class="op">)</span>, <span class="va">methy</span><span class="op">)</span></span></code></pre></div>
<p><img src="epimutacions_files/figure-html/plot_default-1.png" width="700"></p>
<p>If we set the argument <code>gene_annot == TRUE</code> the plot
includes the gene annotations:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_epimutations.html">plot_epimutations</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">epi_mvo</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span><span class="op">)</span>, </span>
<span>                       methy <span class="op">=</span> <span class="va">methy</span>, </span>
<span>                       genes_annot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<p><img src="epimutacions_files/figure-html/plot_genes_annot-1.png" width="700" style="display: block; margin: auto;"></p>
<p>To plot the chromatin marks H3K4me3, H3K27me3 and H3K27ac we need to
specify the argument <code>regulation = TRUE</code>:</p>
<ul>
<li>
<strong>H3K4me3</strong>: commonly associated with the activation of
transcription of nearby genes.</li>
<li>
<strong>H3K27me3</strong>: is used in epigenetics to look for
inactive genes.</li>
<li>
<strong>H3K27ac</strong>: is associated with the higher activation
of transcription and therefore defined as an active enhancer mark</li>
</ul>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_epimutations.html">plot_epimutations</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">epi_mvo</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span><span class="op">)</span>, </span>
<span>                       methy <span class="op">=</span>  <span class="va">methy</span>, </span>
<span>                       regulation <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<p><img src="epimutacions_files/figure-html/plot_genes_regulation-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h2>
<p>The authors would like to thank the team who collaborated in the
initial design of the package in the European BioHackathon 2020:
Lordstrong Akano, James Baye, Alejandro Caceres, Pavlo Hrab, Raquel
Manzano and Margherita Mutarelli. The authors also want to thank the
organization of European BioHackathon 2020 for its support.</p>
<p>All the team members of <em>Project #5</em> for the contribution to
this package:</p>
<table class="table">
<colgroup>
<col width="14%">
<col width="21%">
<col width="17%">
<col width="31%">
<col width="14%">
</colgroup>
<thead><tr class="header">
<th align="left">Name</th>
<th align="left">Surname</th>
<th align="center">ORCID</th>
<th align="left">Affiliation</th>
<th align="left">Team</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Leire</td>
<td align="left">Abarrategui</td>
<td align="center">0000-0002-1175-038X</td>
<td align="left">Faculty of Medical Sciences, Newcastle University,
Newcastle-Upon-Tyne, UK; Autonomous University of Barcelona (UAB),
Barcelona, Spain</td>
<td align="left">Development</td>
</tr>
<tr class="even">
<td align="left">Lordstrong</td>
<td align="left">Akano</td>
<td align="center">0000-0002-1404-0295</td>
<td align="left">College of Medicine, University of Ibadan</td>
<td align="left">Development</td>
</tr>
<tr class="odd">
<td align="left">James</td>
<td align="left">Baye</td>
<td align="center">0000-0002-0078-3688</td>
<td align="left">Wellcome/MRC Cambridge Stem Cell Institute, University
of Cambridge, Cambridge CB2 0AW, UK; Department of Physics, University
of Cambridge, Cambridge CB2 3DY, UK</td>
<td align="left">Development</td>
</tr>
<tr class="even">
<td align="left">Alejandro</td>
<td align="left">Caceres</td>
<td align="center">-</td>
<td align="left">ISGlobal, Barcelona Institute for Global Health, Dr
Aiguader 88, 08003 Barcelona, Spain; Centro de Investigación Biomédica
en Red en Epidemiología y Salud Pública (CIBERESP), Madrid, Spain</td>
<td align="left">Development</td>
</tr>
<tr class="odd">
<td align="left">Carles</td>
<td align="left">Hernandez-Ferrer</td>
<td align="center">0000-0002-8029-7160</td>
<td align="left">Centro Nacional de Análisis Genómico (CNAG-CRG), Center
for Genomic, Regulation; Barcelona Institute of Science and Technology
(BIST), Barcelona, Catalonia, Spain</td>
<td align="left">Development</td>
</tr>
<tr class="even">
<td align="left">Pavlo</td>
<td align="left">Hrab</td>
<td align="center">0000-0002-0742-8478</td>
<td align="left">Department of Genetics and Biotechnology, Biology
faculty, Ivan Franko National University of Lviv</td>
<td align="left">Validation</td>
</tr>
<tr class="odd">
<td align="left">Raquel</td>
<td align="left">Manzano</td>
<td align="center">0000-0002-5124-8992</td>
<td align="left">Cancer Research UK Cambridge Institute; University of
Cambridge, Cambridge, United Kingdom</td>
<td align="left">Reporting</td>
</tr>
<tr class="even">
<td align="left">Margherita</td>
<td align="left">Mutarelli</td>
<td align="center">0000-0002-2168-5059</td>
<td align="left">Institute of Applied Sciences and Intelligent Systems
(ISASI-CNR)</td>
<td align="left">Validation</td>
</tr>
<tr class="odd">
<td align="left">Carlos</td>
<td align="left">Ruiz-Arenas</td>
<td align="center">0000-0002-6014-3498</td>
<td align="left">Centro de Investigación Biomédica en Red de
Enfermedades Raras (CIBERER), Barcelona, Spain; Universitat Pompeu Fabra
(UPF), Barcelona, Spain</td>
<td align="left">Reporting</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code>R version 4.2.1 (2022-06-23 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19044)

Matrix products: default

locale:
[1] LC_COLLATE=Spanish_Spain.utf8  LC_CTYPE=Spanish_Spain.utf8   
[3] LC_MONETARY=Spanish_Spain.utf8 LC_NUMERIC=C                  
[5] LC_TIME=Spanish_Spain.utf8    

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] rtracklayer_1.56.1                                 
 [2] IlluminaHumanMethylationEPICanno.ilm10b2.hg19_0.6.0
 [3] kableExtra_1.3.4                                   
 [4] minfi_1.42.0                                       
 [5] bumphunter_1.38.0                                  
 [6] locfit_1.5-9.6                                     
 [7] iterators_1.0.14                                   
 [8] foreach_1.5.2                                      
 [9] Biostrings_2.64.1                                  
[10] XVector_0.36.0                                     
[11] SummarizedExperiment_1.26.1                        
[12] Biobase_2.56.0                                     
[13] MatrixGenerics_1.8.1                               
[14] matrixStats_0.62.0                                 
[15] GenomicRanges_1.48.0                               
[16] GenomeInfoDb_1.32.4                                
[17] IRanges_2.30.1                                     
[18] S4Vectors_0.34.0                                   
[19] ExperimentHub_2.4.0                                
[20] AnnotationHub_3.4.0                                
[21] BiocFileCache_2.4.0                                
[22] dbplyr_2.2.1                                       
[23] BiocGenerics_0.42.0                                
[24] epimutacions_1.1.2                                 
[25] epimutacionsData_1.0.0                             
[26] BiocStyle_2.24.0                                   

loaded via a namespace (and not attached):
  [1] utf8_1.2.2                             
  [2] tidyselect_1.2.0                       
  [3] htmlwidgets_1.5.4                      
  [4] RSQLite_2.2.18                         
  [5] AnnotationDbi_1.58.0                   
  [6] grid_4.2.1                             
  [7] BiocParallel_1.30.4                    
  [8] munsell_0.5.0                          
  [9] codetools_0.2-18                       
 [10] ragg_1.2.3                             
 [11] preprocessCore_1.58.0                  
 [12] interp_1.1-3                           
 [13] withr_2.5.0                            
 [14] colorspace_2.0-3                       
 [15] OrganismDbi_1.38.1                     
 [16] filelock_1.0.2                         
 [17] highr_0.9                              
 [18] knitr_1.40                             
 [19] rstudioapi_0.14                        
 [20] robustbase_0.95-0                      
 [21] labeling_0.4.2                         
 [22] GenomeInfoDbData_1.2.8                 
 [23] farver_2.1.1                           
 [24] bit64_4.0.5                            
 [25] rhdf5_2.40.0                           
 [26] rprojroot_2.0.3                        
 [27] vctrs_0.4.2                            
 [28] generics_0.1.3                         
 [29] xfun_0.33                              
 [30] biovizBase_1.44.0                      
 [31] R6_2.5.1                               
 [32] illuminaio_0.38.0                      
 [33] AnnotationFilter_1.20.0                
 [34] bitops_1.0-7                           
 [35] rhdf5filters_1.8.0                     
 [36] cachem_1.0.6                           
 [37] reshape_0.8.9                          
 [38] DelayedArray_0.22.0                    
 [39] assertthat_0.2.1                       
 [40] Homo.sapiens_1.3.1                     
 [41] promises_1.2.0.1                       
 [42] BiocIO_1.6.0                           
 [43] scales_1.2.1                           
 [44] nnet_7.3-18                            
 [45] gtable_0.3.1                           
 [46] ensembldb_2.20.2                       
 [47] rlang_1.0.6                            
 [48] genefilter_1.78.0                      
 [49] systemfonts_1.0.4                      
 [50] splines_4.2.1                          
 [51] lazyeval_0.2.2                         
 [52] GEOquery_2.64.2                        
 [53] dichromat_2.0-0.1                      
 [54] checkmate_2.1.0                        
 [55] BiocManager_1.30.18                    
 [56] yaml_2.3.5                             
 [57] reshape2_1.4.4                         
 [58] backports_1.4.1                        
 [59] GenomicFeatures_1.48.4                 
 [60] httpuv_1.6.6                           
 [61] Hmisc_4.7-1                            
 [62] RBGL_1.72.0                            
 [63] tools_4.2.1                            
 [64] bookdown_0.29                          
 [65] nor1mix_1.3-0                          
 [66] ggplot2_3.3.6                          
 [67] ellipsis_0.3.2                         
 [68] jquerylib_0.1.4                        
 [69] RColorBrewer_1.1-3                     
 [70] siggenes_1.70.0                        
 [71] Rcpp_1.0.9                             
 [72] plyr_1.8.7                             
 [73] base64enc_0.1-3                        
 [74] sparseMatrixStats_1.8.0                
 [75] progress_1.2.2                         
 [76] zlibbioc_1.42.0                        
 [77] purrr_0.3.5                            
 [78] RCurl_1.98-1.9                         
 [79] prettyunits_1.1.1                      
 [80] rpart_4.1.16                           
 [81] deldir_1.0-6                           
 [82] openssl_2.0.4                          
 [83] cluster_2.1.4                          
 [84] ggrepel_0.9.1                          
 [85] fs_1.5.2                               
 [86] magrittr_2.0.3                         
 [87] data.table_1.14.4                      
 [88] ProtGenerics_1.28.0                    
 [89] hms_1.1.2                              
 [90] mime_0.12                              
 [91] evaluate_0.17                          
 [92] xtable_1.8-4                           
 [93] XML_3.99-0.11                          
 [94] jpeg_0.1-9                             
 [95] mclust_5.4.10                          
 [96] gridExtra_2.3                          
 [97] compiler_4.2.1                         
 [98] biomaRt_2.52.0                         
 [99] tibble_3.1.8                           
[100] crayon_1.5.2                           
[101] htmltools_0.5.3                        
[102] later_1.3.0                            
[103] tzdb_0.3.0                             
[104] Formula_1.2-4                          
[105] tidyr_1.2.1                            
[106] DBI_1.1.3                              
[107] MASS_7.3-58.1                          
[108] rappdirs_0.3.3                         
[109] Matrix_1.5-1                           
[110] readr_2.1.3                            
[111] cli_3.4.1                              
[112] quadprog_1.5-8                         
[113] Gviz_1.40.1                            
[114] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2
[115] pkgconfig_2.0.3                        
[116] pkgdown_2.0.6                          
[117] GenomicAlignments_1.32.1               
[118] foreign_0.8-83                         
[119] xml2_1.3.3                             
[120] svglite_2.1.0                          
[121] annotate_1.74.0                        
[122] bslib_0.4.0                            
[123] rngtools_1.5.2                         
[124] multtest_2.52.0                        
[125] beanplot_1.3.1                         
[126] webshot_0.5.4                          
[127] rvest_1.0.3                            
[128] doRNG_1.8.2                            
[129] scrime_1.3.5                           
[130] VariantAnnotation_1.42.1               
[131] stringr_1.4.1                          
[132] digest_0.6.29                          
[133] graph_1.74.0                           
[134] rmarkdown_2.17                         
[135] base64_2.0.1                           
[136] htmlTable_2.4.1                        
[137] DelayedMatrixStats_1.18.2              
[138] restfulr_0.0.15                        
[139] curl_4.3.3                             
[140] shiny_1.7.2                            
[141] Rsamtools_2.12.0                       
[142] rjson_0.2.21                           
[143] lifecycle_1.0.3                        
[144] nlme_3.1-160                           
[145] jsonlite_1.8.2                         
[146] Rhdf5lib_1.18.2                        
[147] BSgenome_1.64.0                        
[148] desc_1.4.2                             
[149] viridisLite_0.4.1                      
[150] askpass_1.1                            
[151] limma_3.52.4                           
[152] fansi_1.0.3                            
[153] pillar_1.8.1                           
[154] lattice_0.20-45                        
[155] GO.db_3.15.0                           
[156] KEGGREST_1.36.3                        
[157] fastmap_1.1.0                          
[158] httr_1.4.4                             
[159] DEoptimR_1.0-11                        
[160] survival_3.4-0                         
[161] interactiveDisplayBase_1.34.0          
[162] glue_1.6.2                             
[163] png_0.1-7                              
[164] isotree_0.5.17                         
[165] BiocVersion_3.15.2                     
[166] bit_4.0.4                              
[167] stringi_1.7.8                          
[168] sass_0.4.2                             
[169] HDF5Array_1.24.2                       
[170] blob_1.2.3                             
[171] textshaping_0.3.6                      
[172] org.Hs.eg.db_3.15.0                    
[173] latticeExtra_0.6-30                    
[174] memoise_2.0.1                          
[175] dplyr_1.0.10                           </code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-aref2019" class="csl-entry">
Aref-Eshghi, Erfan, Eric G. Bend, Samantha Colaiacovo, Michelle Caudle,
Rana Chakrabarti, Melanie Napier, Lauren Brick, et al. 2019.
<span>“Diagnostic Utility of Genome-Wide DNA Methylation Testing in
Genetically Unsolved Individuals with Suspected Hereditary
Conditions.”</span> <em>The American Journal of Human Genetics</em>.
https://doi.org/<a href="https://doi.org/10.1016/j.ajhg.2019.03.008" class="external-link">https://doi.org/10.1016/j.ajhg.2019.03.008</a>.
</div>
<div id="ref-barbosa2018identification" class="csl-entry">
Barbosa, Mafalda, Ricky S Joshi, Paras Garg, Alejandro Martin-Trujillo,
Nihir Patel, Bharati Jadhav, Corey T Watson, et al. 2018.
<span>“Identification of Rare de Novo Epigenetic Variations in
Congenital Disorders.”</span> <em>Nature Communications</em> 9 (1):
1–11.
</div>
<div id="ref-jaffe2012bump" class="csl-entry">
Jaffe, Andrew E, Peter Murakami, Hwajin Lee, Jeffrey T Leek, M Daniele
Fallin, Andrew P Feinberg, and Rafael A Irizarry. 2012. <span>“Bump
Hunting to Identify Differentially Methylated Regions in Epigenetic
Epidemiology Studies.”</span> <em>International Journal of
Epidemiology</em> 41 (1): 200–209.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Dolors Pelegri-Siso, Juan R. Gonzalez, Carlos Ruiz-Arenas, Carles Hernandez-Ferrer, Leire Abarrategui.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
